File: 20-actors/watatsuna/methods/analyze.py
#!/usr/bin/env python3
"""watatsuna 綿津綱 — submarine-cable network resilience analyzer.

ADR-2606012600. Reads a kotoba-EDN cable graph (:cable/* systems, :station/*
landing stations, :cable.link/* incidence, :cable.seg/* chokepoint-traversing
segments, :cable.fault/* observed bulletins) and emits:

  1. an AGGREGATE-FIRST resilience report (out/intel-report.md) — where the
     world's submarine capacity concentrates onto a chokepoint / single point of
     failure, framed toward redundancy + diversity-routing + faster repair.
  2. the derived resilience datoms (out/cable-criticality.kotoba.edn), flagged
     :derived — never re-ingested as authoritative fact.

CONSTITUTIONAL framing (watatsumi N8 + Charter Rider §2(d)): this is a
RESILIENCE map, NEVER a target-list. Output ranks chokepoints by fragility so
the network can be made MORE robust (add diverse routes, pre-stage repair
ships) — it does NOT identify where to cut. Sabotage intent is never asserted;
fault :kind mirrors only the public bulletin's own classification.

stdlib only (no numpy). Usage:
    python3 analyze.py [seed.edn] [--out OUTDIR]
"""
from __future__ import annotations
import sys, re, pathlib
from collections import defaultdict

# ── minimal EDN reader (subset: [] {} :kw "str" num bool nil) — ported from tsumugi
_TOK = re.compile(r'[\s,]+|;[^\n]*|(\[|\]|\{|\}|"(?:\\.|[^"\\])*"|[^\s,\[\]{}]+)')
_END = object()


def _tokens(s: str):
    for m in _TOK.finditer(s):
        t = m.group(1)
        if t is not None:
            yield t


def _atom(t: str):
    if t.startswith('"'):
        return t[1:-1].replace('\\"', '"').replace('\\\\', '\\')
    if t == 'true':
        return True
    if t == 'false':
        return False
    if t == 'nil':
        return None
    if t.startswith(':'):
        return t  # keep keywords as ":ns/name" strings
    try:
        return int(t)
    except ValueError:
        try:
            return float(t)
        except ValueError:
            return t


def _parse(it):
    t = next(it)
    if t == '[':
        out = []
        while (x := _parse(it)) is not _END:
            out.append(x)
        return out
    if t == '{':
        out = {}
        while (k := _parse(it)) is not _END:
            v = _parse(it)
            out[k] = v
        return out
    if t in (']', '}'):
        return _END
    return _atom(t)


def load_edn(path: pathlib.Path):
    it = _tokens(path.read_text(encoding='utf-8'))
    return _parse(it)


# ── classify the flat datom vector into entity buckets
def classify(rows):
    cables, stations, links, segs, faults = {}, {}, [], [], []
    for r in rows:
        if not isinstance(r, dict):
            continue
        if ':cable/id' in r:
            cables[r[':cable/id']] = r
        elif ':station/id' in r:
            stations[r[':station/id']] = r
        elif ':cable.link/id' in r:
            links.append(r)
        elif ':cable.seg/id' in r:
            segs.append(r)
        elif ':cable.fault/id' in r:
            faults.append(r)
    return cables, stations, links, segs, faults


def analyze(cables, stations, links, segs, faults):
    cap = {c: cables[c].get(':cable/design-capacity-tbps', 0.0) for c in cables}

    # per-station: which cables land here
    station_cables = defaultdict(set)
    for lk in links:
        station_cables[lk[':cable.link/station']].add(lk[':cable.link/cable'])

    station_degree = {s: len(station_cables[s]) for s in stations}
    station_capacity = {
        s: round(sum(cap.get(c, 0.0) for c in station_cables[s]), 2) for s in stations
    }

    # per-chokepoint: distinct cables traversing it (via segments) + via station tags
    choke_cables = defaultdict(set)
    for sg in segs:
        for cp in (sg.get(':cable.seg/traverses') or []):
            choke_cables[cp].add(sg[':cable.seg/cable'])
    # also fold station chokepoint tags: a cable landing behind a chokepoint depends on it
    for s, meta in stations.items():
        for cp in (meta.get(':station/chokepoint') or []):
            for c in station_cables[s]:
                choke_cables[cp].add(c)

    choke_load = {
        cp: round(sum(cap.get(c, 0.0) for c in cs), 2) for cp, cs in choke_cables.items()
    }
    choke_count = {cp: len(cs) for cp, cs in choke_cables.items()}

    # per-cable diversity: # distinct chokepoints it depends on (lower = more brittle)
    cable_chokes = defaultdict(set)
    for cp, cs in choke_cables.items():
        for c in cs:
            cable_chokes[c].add(cp)
    cable_diversity = {c: len(cable_chokes[c]) for c in cables}

    # redundancy gap: landing stations served by a single cable
    redundancy_gap = sorted(s for s in stations if station_degree[s] <= 1)

    return dict(
        cap=cap,
        station_cables=station_cables,
        station_degree=station_degree,
        station_capacity=station_capacity,
        choke_cables=choke_cables,
        choke_load=choke_load,
        choke_count=choke_count,
        cable_diversity=cable_diversity,
        redundancy_gap=redundancy_gap,
    )


def name(d, key, idkey, namekey):
    return d.get(key, {}).get(namekey, key) if isinstance(d.get(key), dict) else key


def render_report(cables, stations, links, segs, faults, a):
    L = []
    P = L.append
    P("# watatsuna 綿津綱 — submarine-cable network resilience report")
    P("")
    P("> ADR-2606012600 · **aggregate-first** · RESILIENCE map (NOT a target-list; "
      "watatsumi N8 + Charter Rider §2(d)). Sabotage intent never asserted. "
      "All sourcing `:representative` — bounded illustrative seed, not exhaustive coverage.")
    P("")
    P(f"- cable systems: **{len(cables)}**  ·  landing stations: **{len(stations)}**  "
      f"·  cable↔station links: **{len(links)}**  ·  chokepoint segments: **{len(segs)}**  "
      f"·  observed faults: **{len(faults)}**")
    total_design = round(sum(a['cap'].values()), 1)
    P(f"- total seeded design capacity: **{total_design} Tbps**")
    P("")

    # ── chokepoint concentration (the headline resilience signal) ──
    P("## Chokepoint concentration — single-point-of-failure surface")
    P("")
    P("Capacity (Σ design Tbps of distinct cables) that depends on each maritime "
      "chokepoint. Higher = more of the world's traffic shares one geographic fate. "
      "**Routed to redundancy + pre-staged repair, never to interdiction.**")
    P("")
    P("| chokepoint | cables | dependent capacity (Tbps) |")
    P("|---|---:|---:|")
    for cp in sorted(a['choke_load'], key=lambda k: -a['choke_load'][k]):
        P(f"| `{cp}` | {a['choke_count'][cp]} | {a['choke_load'][cp]} |")
    P("")

    # ── landing-station hubs ──
    P("## Landing-station hubs — convergence points")
    P("")
    P("| station | country | cables | landed capacity (Tbps) |")
    P("|---|---|---:|---:|")
    for s in sorted(stations, key=lambda k: (-a['station_degree'][k], -a['station_capacity'][k])):
        if a['station_degree'][s] == 0:
            continue
        st = stations[s]
        P(f"| {st.get(':station/name', s)} | {st.get(':station/country','?')} "
          f"| {a['station_degree'][s]} | {a['station_capacity'][s]} |")
    P("")

    # ── brittle cables (low route diversity) ──
    P("## Most brittle systems — low chokepoint diversity")
    P("")
    P("Cables depending on the fewest distinct chokepoints carry the highest "
      "correlated-failure risk; candidates for diverse-route augmentation.")
    P("")
    P("| cable | chokepoints depended on | design (Tbps) | RFS |")
    P("|---|---:|---:|---:|")
    for c in sorted(cables, key=lambda k: (a['cable_diversity'][k], -a['cap'][k])):
        div = a['cable_diversity'][c]
        if div == 0:
            continue  # no charted chokepoint dependency in seed
        cb = cables[c]
        P(f"| {cb.get(':cable/name', c)} | {div} | {a['cap'][c]} | "
          f"{cb.get(':cable/rfs-year','?')} |")
    P("")

    # ── redundancy gaps ──
    P("## Redundancy gaps — single-cable landing stations")
    P("")
    if a['redundancy_gap']:
        for s in a['redundancy_gap']:
            P(f"- {stations[s].get(':station/name', s)} "
              f"({stations[s].get(':station/country','?')}) — routed to redundancy")
    else:
        P("- (none in seed)")
    P("")

    # ── observed faults (as-of history) ──
    P("## Observed fault bulletins (Datom as-of history; 非終末論 — appended, never overwritten)")
    P("")
    P("| fault | cable | kind (bulletin's own) | detected | restored |")
    P("|---|---|---|---|---|")
    for f in faults:
        cb = cables.get(f.get(':cable.fault/cable'), {})
        P(f"| `{f[':cable.fault/id']}` | {cb.get(':cable/name', f.get(':cable.fault/cable'))} "
          f"| `{f.get(':cable.fault/kind')}` | {f.get(':cable.fault/detected-at','?')} "
          f"| {f.get(':cable.fault/restored-at') or 'open'} |")
    P("")
    P("---")
    P("*Generated by `watatsuna/methods/analyze.py`. HONEST: R0 bounded seed; "
      "coordinates rounded to landing town; capacities are public design figures; "
      "chokepoint dependency is charted only for seeded segments. Live planet-scale "
      "ingest is G7 Council+operator gated.*")
    return "\n".join(L) + "\n"


def render_datoms(cables, stations, a):
    L = []
    P = L.append
    P(";; watatsuna — DERIVED resilience datoms (ADR-2606012600). :derived — NOT authoritative fact.")
    P(";; Recomputed from the seed graph; do not re-ingest as :authoritative.")
    P("[")
    for cp in sorted(a['choke_load'], key=lambda k: -a['choke_load'][k]):
        P(f' {{:resilience/chokepoint "{cp}" :resilience/chokepoint-load {a["choke_load"][cp]} '
          f':resilience/cable-count {a["choke_count"][cp]} :resilience/derived true}}')
    for s in sorted(stations, key=lambda k: -a['station_degree'][k]):
        if a['station_degree'][s] == 0:
            continue
        P(f' {{:resilience/station "{s}" :resilience/station-degree {a["station_degree"][s]} '
          f':resilience/station-capacity-tbps {a["station_capacity"][s]} :resilience/derived true}}')
    for c in sorted(cables, key=lambda k: a['cable_diversity'][k]):
        P(f' {{:resilience/cable "{c}" :resilience/cable-diversity {a["cable_diversity"][c]} '
          f':resilience/derived true}}')
    P("]")
    return "\n".join(L) + "\n"


def main(argv):
    here = pathlib.Path(__file__).resolve().parent.parent
    seed = pathlib.Path(argv[1]) if len(argv) > 1 and not argv[1].startswith('--') \
        else here / "data" / "seed-cable-graph.kotoba.edn"
    outdir = here / "out"
    if "--out" in argv:
        outdir = pathlib.Path(argv[argv.index("--out") + 1])
    outdir.mkdir(parents=True, exist_ok=True)

    rows = load_edn(seed)
    cables, stations, links, segs, faults = classify(rows)
    a = analyze(cables, stations, links, segs, faults)

    (outdir / "intel-report.md").write_text(
        render_report(cables, stations, links, segs, faults, a), encoding='utf-8')
    (outdir / "cable-criticality.kotoba.edn").write_text(
        render_datoms(cables, stations, a), encoding='utf-8')

    print(f"watatsuna: {len(cables)} cables, {len(stations)} stations, "
          f"{len(links)} links, {len(segs)} segments, {len(faults)} faults")
    top = sorted(a['choke_load'], key=lambda k: -a['choke_load'][k])[:3]
    print("top chokepoints by dependent capacity: " +
          ", ".join(f"{cp} {a['choke_load'][cp]}Tbps" for cp in top))
    print(f"wrote {outdir/'intel-report.md'} + {outdir/'cable-criticality.kotoba.edn'}")


if __name__ == "__main__":
    main(sys.argv)

File: 20-actors/watatsuna/methods/analyze.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/analyze.py (replaces the failed unit_refactor cljc stub).
(ns watatsuna.methods.analyze
  "watatsuna 綿津綱 — submarine-cable network resilience analyzer (ADR-2606012600).

  Reads a kotoba-EDN cable graph (:cable/* systems, :station/* landing stations, :cable.link/*
  incidence, :cable.seg/* chokepoint-traversing segments, :cable.fault/* observed bulletins)
  and emits an AGGREGATE-FIRST resilience report + derived :resilience/* datoms (flagged
  :derived, never re-ingested as authoritative fact).

  CONSTITUTIONAL (watatsumi N8 + Charter Rider §2(d)): a RESILIENCE map, NEVER a target-list.
  Chokepoints are ranked by fragility so the network can be made MORE robust (diverse routes,
  pre-staged repair) — it does NOT identify where to cut. Sabotage intent is never asserted;
  fault :kind mirrors only the public bulletin's own classification (the Python tokenizer is
  replaced by clojure.edn — keywords read natively).

  Run:  bb --classpath 20-actors 20-actors/watatsuna/methods/analyze.clj"
  (:require [clojure.java.io :as io]
            [clojure.edn :as edn]
            [clojure.string :as str]))

(def ^:private this-file *file*)
(defn- actor-root [] (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile))
(defn- round2 [x] (/ (Math/round (* (double x) 100.0)) 100.0))

(defn load-edn [path] (edn/read-string (slurp (io/file path))))

(defn classify
  "Bucket a flat datom vector into cable/station/link/seg/fault entities."
  [rows]
  (reduce
   (fn [out r]
     (if-not (map? r)
       out
       (cond
         (:cable/id r)       (assoc-in out [:cables (:cable/id r)] r)
         (:station/id r)     (assoc-in out [:stations (:station/id r)] r)
         (:cable.link/id r)  (update out :links conj r)
         (:cable.seg/id r)   (update out :segs conj r)
         (:cable.fault/id r) (update out :faults conj r)
         :else out)))
   {:cables {} :stations {} :links [] :segs [] :faults []}
   rows))

(defn analyze [{:keys [cables stations links segs]}]
  (let [cap (into {} (map (fn [[id c]] [id (or (:cable/design-capacity-tbps c) 0.0)]) cables))
        ;; per-station: which cables land here
        station-cables (reduce (fn [m lk]
                                 (update m (:cable.link/station lk) (fnil conj #{}) (:cable.link/cable lk)))
                               {} links)
        sc (fn [s] (get station-cables s #{}))
        station-degree (into {} (map (fn [s] [s (count (sc s))]) (keys stations)))
        station-capacity (into {} (map (fn [s] [s (round2 (reduce + 0.0 (map #(get cap % 0.0) (sc s))))])
                                       (keys stations)))
        ;; per-chokepoint: distinct cables traversing it (via segments) + via station tags
        choke-cables (as-> {} m
                       (reduce (fn [m sg]
                                 (reduce (fn [m cp] (update m cp (fnil conj #{}) (:cable.seg/cable sg)))
                                         m (or (:cable.seg/traverses sg) [])))
                               m segs)
                       (reduce (fn [m [s meta]]
                                 (reduce (fn [m cp]
                                           (reduce (fn [m c] (update m cp (fnil conj #{}) c))
                                                   m (sc s)))
                                         m (or (:station/chokepoint meta) [])))
                               m stations))
        choke-load (into {} (map (fn [[cp cs]] [cp (round2 (reduce + 0.0 (map #(get cap % 0.0) cs)))]) choke-cables))
        choke-count (into {} (map (fn [[cp cs]] [cp (count cs)]) choke-cables))
        ;; per-cable diversity: # distinct chokepoints it depends on (lower = more brittle)
        cable-chokes (reduce (fn [m [cp cs]] (reduce (fn [m c] (update m c (fnil conj #{}) cp)) m cs))
                             {} choke-cables)
        cable-diversity (into {} (map (fn [c] [c (count (get cable-chokes c #{}))]) (keys cables)))
        redundancy-gap (sort (filter #(<= (station-degree %) 1) (keys stations)))]
    {:cap cap :station-cables station-cables :station-degree station-degree
     :station-capacity station-capacity :choke-cables choke-cables :choke-load choke-load
     :choke-count choke-count :cable-diversity cable-diversity :redundancy-gap redundancy-gap}))

(defn render-report [{:keys [cables stations links segs faults]} a]
  (let [P (fn [& xs] (apply str xs))
        total-design (round2 (reduce + 0.0 (vals (:cap a))))]
    (str/join
     "\n"
     (concat
      ["# watatsuna 綿津綱 — submarine-cable network resilience report"
       ""
       (str "> ADR-2606012600 · **aggregate-first** · RESILIENCE map (NOT a target-list; "
            "watatsumi N8 + Charter Rider §2(d)). Sabotage intent never asserted. "
            "All sourcing `:representative` — bounded illustrative seed, not exhaustive coverage.")
       ""
       (P "- cable systems: **" (count cables) "**  ·  landing stations: **" (count stations)
          "**  ·  cable↔station links: **" (count links) "**  ·  chokepoint segments: **"
          (count segs) "**  ·  observed faults: **" (count faults) "**")
       (P "- total seeded design capacity: **" total-design " Tbps**")
       ""
       "## Chokepoint concentration — single-point-of-failure surface"
       ""
       (str "Capacity (Σ design Tbps of distinct cables) that depends on each maritime "
            "chokepoint. Higher = more of the world's traffic shares one geographic fate. "
            "**Routed to redundancy + pre-staged repair, never to interdiction.**")
       ""
       "| chokepoint | cables | dependent capacity (Tbps) |"
       "|---|---:|---:|"]
      (for [cp (sort-by (juxt #(- (get-in a [:choke-load %])) str) (keys (:choke-load a)))]
        (P "| `" cp "` | " (get-in a [:choke-count cp]) " | " (get-in a [:choke-load cp]) " |"))
      ["" "## Landing-station hubs — convergence points" ""
       "| station | country | cables | landed capacity (Tbps) |"
       "|---|---|---:|---:|"]
      (for [s (sort-by (juxt #(- (get-in a [:station-degree %]))
                             #(- (get-in a [:station-capacity %])) str) (keys stations))
            :when (pos? (get-in a [:station-degree s]))]
        (P "| " (get-in stations [s :station/name] s) " | " (get-in stations [s :station/country] "?")
           " | " (get-in a [:station-degree s]) " | " (get-in a [:station-capacity s]) " |"))
      ["" "## Most brittle systems — low chokepoint diversity" ""
       (str "Cables depending on the fewest distinct chokepoints carry the highest "
            "correlated-failure risk; candidates for diverse-route augmentation.")
       ""
       "| cable | chokepoints depended on | design (Tbps) | RFS |"
       "|---|---:|---:|---:|"]
      (for [c (sort-by (juxt #(get-in a [:cable-diversity %]) #(- (get-in a [:cap %])) str) (keys cables))
            :when (pos? (get-in a [:cable-diversity c]))]
        (P "| " (get-in cables [c :cable/name] c) " | " (get-in a [:cable-diversity c]) " | "
           (get-in a [:cap c]) " | " (get-in cables [c :cable/rfs-year] "?") " |"))
      ["" "## Redundancy gaps — single-cable landing stations" ""]
      (if (seq (:redundancy-gap a))
        (for [s (:redundancy-gap a)]
          (P "- " (get-in stations [s :station/name] s) " ("
             (get-in stations [s :station/country] "?") ") — routed to redundancy"))
        ["- (none in seed)"])
      ["" "## Observed fault bulletins (Datom as-of history; 非終末論 — appended, never overwritten)" ""
       "| fault | cable | kind (bulletin's own) | detected | restored |"
       "|---|---|---|---|---|"]
      (for [f faults]
        (P "| `" (:cable.fault/id f) "` | "
           (get-in cables [(:cable.fault/cable f) :cable/name] (:cable.fault/cable f))
           " | `" (:cable.fault/kind f) "` | " (or (:cable.fault/detected-at f) "?")
           " | " (or (:cable.fault/restored-at f) "open") " |"))
      ["" "---"
       (str "*Generated by `watatsuna/methods/analyze.clj`. HONEST: R0 bounded seed; coordinates "
            "rounded to landing town; capacities are public design figures; chokepoint dependency "
            "is charted only for seeded segments. Live planet-scale ingest is G7 Council+operator gated.*")
       ""]))))

(defn render-datoms [{:keys [cables stations]} a]
  (str/join
   "\n"
   (concat
    [";; watatsuna — DERIVED resilience datoms (ADR-2606012600). :derived — NOT authoritative fact."
     ";; Recomputed from the seed graph; do not re-ingest as :authoritative."
     "["]
    (for [cp (sort-by (juxt #(- (get-in a [:choke-load %])) str) (keys (:choke-load a)))]
      (format " {:resilience/chokepoint %s :resilience/chokepoint-load %s :resilience/cable-count %d :resilience/derived true}"
              (pr-str cp) (get-in a [:choke-load cp]) (get-in a [:choke-count cp])))
    (for [s (sort-by (juxt #(- (get-in a [:station-degree %])) str) (keys stations))
          :when (pos? (get-in a [:station-degree s]))]
      (format " {:resilience/station %s :resilience/station-degree %d :resilience/station-capacity-tbps %s :resilience/derived true}"
              (pr-str s) (get-in a [:station-degree s]) (get-in a [:station-capacity s])))
    (for [c (sort-by (juxt #(get-in a [:cable-diversity %]) str) (keys cables))]
      (format " {:resilience/cable %s :resilience/cable-diversity %d :resilience/derived true}"
              (pr-str c) (get-in a [:cable-diversity c])))
    ["]" ""])))

(defn main [& argv]
  (let [args (vec argv)
        out-idx (.indexOf args "--out")
        out-val (when (>= out-idx 0) (nth args (inc out-idx)))
        out (if out-val (io/file out-val) (io/file (actor-root) "out"))
        seed (or (first (remove #(or (str/starts-with? % "--") (= % out-val)) args))
                 (str (io/file (actor-root) "data" "seed-cable-graph.kotoba.edn")))
        g (classify (load-edn seed))
        a (analyze g)]
    (.mkdirs out)
    (spit (io/file out "intel-report.md") (render-report g a))
    (spit (io/file out "cable-criticality.kotoba.edn") (render-datoms g a))
    (println (format "watatsuna: %d cables, %d stations, %d links, %d segments, %d faults"
                     (count (:cables g)) (count (:stations g)) (count (:links g))
                     (count (:segs g)) (count (:faults g))))
    (let [top (take 3 (sort-by (comp - val) (:choke-load a)))]
      (println (str "top chokepoints by dependent capacity: "
                    (str/join ", " (map (fn [[cp l]] (str cp " " l "Tbps")) top)))))))

(when (= *file* (System/getProperty "babashka.file"))
  (apply main *command-line-args*))
